File: Atividade5/BANCO-OO/src/classes/Banco.java
package classes;

import static utilitarios.Cadastro.*;
import static utilitarios.Utils.*;
import static utilitarios.EntradaDeDados.*;

import java.util.ArrayList;
import java.util.List;           

public class Banco {
    private List<Cliente> clientes = inicializarClientes();
    private List<Conta> contas = inicializarContas(clientes);

    // Inserir
    public void associarConta(){
        Cliente cliente = null;
        Conta conta = null;
        int opcao = recebeInt("Já é cliente do banco? (1-sim/2-não): ");
        limparBuffer();

        if(opcao == 1){
            String cpf = recebeStr("Digite seu CPF:");
            cliente = procurarPeloCPF(cpf);
            if(cliente != null){
                System.out.println("\nCliente encontrado(a)!\n" + cliente.dados());
                conta = criarConta(cliente, this.contas);
                
            }else{
                System.out.println("Cliente não encontrado.");
                return;
            }
        }
        else{
            cliente = cadastrarCliente(this.clientes);
            conta = criarConta(cliente, this.contas);
            this.clientes.add(cliente);
        }
        msgSucesso();
        this.contas.add(conta);
        cliente.adicionarConta(conta);
    }

    // Consultar
    public void consultarPeloNumero(){
        String numero = recebeStr("Número da conta:");

        for(Conta conta : this.contas){
            if(numero.equals(conta.getNumero())){
                System.out.println("Conta encontrada!");
                conta.mostrarInfo();
                return;
            }
        }
        msgContaInexistente();
    }

    public Cliente procurarPeloCPF(String cpf){
    
        for(Cliente cliente : this.clientes){
            if(cpf.equals(cliente.getCPF())) {
                return cliente;
            }
        }
       
        return null;
    }

    public Conta procurarPeloNumero(String numero){

        for(Conta conta : this.contas){
            if(numero.equals(conta.getNumero())){
                return conta;
            }
        }

        return null;
    }

    // atualizar
    public void atualizarNomeTitular(){
        String numeroConta = recebeStr("Número da conta que quer atualizar:");
        Conta contaProcurada = procurarPeloNumero(numeroConta);

        if (contaProcurada != null) {
            System.out.println("Conta encontrada!\nDados do Titular:\n");
            Cliente clienteConta = contaProcurada.getTitular();
            System.out.println(clienteConta.dados());
            String nome = recebeStr("-> Novo nome do titular:");
            clienteConta.setNome(nome);
            msgSucesso();
            return;
        }  
        msgContaInexistente();
    }
    
    // excluir
    public void excluirConta(){
        String numeroConta = recebeStr("Número da conta que quer excluir:");
        Conta contaProcurada = procurarPeloNumero(numeroConta);

        if(contaProcurada != null){
            this.contas.remove(contaProcurada);
            Cliente cliente = contaProcurada.getTitular();
            cliente.getContas().remove(contaProcurada);
            msgSucesso();
            return;
        }
        msgContaInexistente();
    }

    public void excluirCliente(){
        String cpf = recebeStr("Digite seu CPF:");
        Cliente cliente = procurarPeloCPF(cpf);

        if(cliente != null){
           
            for(Conta conta : cliente.getContas()){
                this.contas.remove(conta);
            }

            this.clientes.remove(cliente);
            msgSucesso();
            return;
        }
        System.out.println("Cliente não encontrado.");
    }

    public void fazerSaque(){
        String numeroConta = recebeStr("Número da conta:");
        Conta contaProcurada = procurarPeloNumero(numeroConta);

        if(contaProcurada != null){
            double valor = recebeValor("Valor do saque:");
            contaProcurada.sacar(valor);
            return;
        }
        msgContaInexistente();
    }

    public void fazerDeposito(){
        String numeroConta = recebeStr("Número da conta:");
        Conta contaProcurada = procurarPeloNumero(numeroConta);

        if(contaProcurada != null){
            contaProcurada.depositar( recebeValor("Valor do Depósito:"));
            return;
        }
        msgContaInexistente();
    }

    public void fazerTransferencia(){
        String numeroConta = recebeStr("Número da conta:");
        Conta contaProcurada = procurarPeloNumero(numeroConta);

        if(contaProcurada != null){
            Conta contaDestino = procurarPeloNumero(recebeStr("Número da conta que quer transferir:"));
            if(contaDestino != null){
                double valor = recebeValor("Valor da tranferência:");
                contaProcurada.transferir(contaDestino, valor);
                msgSucesso();
            }
            else msgContaInexistente();
        }
        else msgContaInexistente();
    }

    public List<Conta> receberContasDestinoTranferencia(int qtdContas){
        List<Conta> contasDestino = new ArrayList<>();

        for(int i = 0; i < qtdContas; i++){
            String numeroConta = recebeStr("Número da conta:");
            Conta contaProcurada = procurarPeloNumero(numeroConta);

            while(contaProcurada == null){
                System.out.println("\nConta não encontrada. Tente novamente:");
                numeroConta = recebeStr("Número da conta:");
                contaProcurada = procurarPeloNumero(numeroConta);
            }

            contasDestino.add(contaProcurada);
        }

        return contasDestino;
    }

    public void TransferenciaVariasContas(){
        String numeroConta = recebeStr("Número da conta:");
        Conta contaProcurada = procurarPeloNumero(numeroConta);

        if(contaProcurada != null){
            int qtdContas = recebeInt("Quantidade de contas que quer tranferir:");
            limparBuffer();
            List<Conta> contasDestino = receberContasDestinoTranferencia(qtdContas);

            for(Conta conta : contasDestino){
                if(contaProcurada.getSaldo() == 0){
                    System.out.println("\nSaldo insuficiente para continuar as tranferências.");
                    return;
                }

                double valor = recebeValor("Valor da transferência para conta " + conta.getNumero() + ":");

                while(valor > contaProcurada.getSaldo()){
                    System.out.println("Valor acima do saldo. Tente novamente:");
                    valor = recebeValor("Valor da transferência para conta " + conta.getNumero() + ":");

                }
                contaProcurada.transferir(conta, valor);
            }
            msgSucesso();
            return;
        }
        msgContaInexistente();
    }

    public void mudarTitularidade(){
        String numeroConta = recebeStr("Número da conta:");
        Conta contaProcurada = procurarPeloNumero(numeroConta);

        if(contaProcurada != null){
            String cpf = recebeStr("Digite o CPF do novo titular:");
            Cliente clienteNovo = procurarPeloCPF(cpf);

            if(clienteNovo == null){       // novo titular não é cliente do banco
                System.out.println("Cliente não cadastrado. Fazer cadastro: ");
                clienteNovo = cadastrarCliente(this.clientes);
                this.clientes.add(clienteNovo);
            }
            
            Cliente clienteAntigo = contaProcurada.getTitular();
            clienteAntigo.getContas().remove(contaProcurada);

            contaProcurada.setTitular(clienteNovo);
            clienteNovo.adicionarConta(contaProcurada);
            msgSucesso();
        }
        else{
            msgContaInexistente();
        }

    }

    public void somarSaldoContasCliente(){
        double somaSaldo = 0;
        String cpf = recebeStr("CPF: ");
        Cliente cliente = procurarPeloCPF(cpf);

        if(cliente != null){
            for(Conta conta : cliente.getContas()){
                somaSaldo += conta.getSaldo();
                conta.mostrarInfo();
            }

            System.out.printf("\nSaldo total de todas as contas do cliente %s: R$ %.2f", cliente.getNome(), somaSaldo);
            return;
        }

        msgContaInexistente();
    }

    public void mostrarQtdContas(){
        System.out.printf("\n> QUANTIDADE DE CONTAS DO BANCO: %d", this.contas.size());
    }

    public double somarSaldoContasBanco(){
        double saldoTotal = 0;

        for (Conta conta : this.contas){
            saldoTotal += conta.getSaldo();
        }

        return saldoTotal;
    }

    public void mostrarSaldoTotalContasBanco(){
        System.out.printf("\n> SALDO TOTAL DE TODAS AS CONTAS DO BANCO: %.2f", somarSaldoContasBanco());

    }

    public void mostrarMediaSaldoContasBanco(){
        double saldoTotal = somarSaldoContasBanco();
        double mediaSaldo = saldoTotal / this.contas.size();

        System.out.printf("> MÉDIA DO SALDO DAS CONTAS DO BANCO: %.2f", mediaSaldo);
    }

    public void listarContasCliente(){
        String cpf = recebeStr("CPF: ");
        Cliente cliente = procurarPeloCPF(cpf);

        if(cliente != null){
            System.out.printf("---- CONTAS DO(A) CLIENTE %s ----\n", cliente.getNome().toUpperCase());
            for(Conta conta : cliente.getContas()){
                conta.mostrarInfo();
            }
            return;
        }
        msgContaInexistente();
    }

    public void listarContas(){
        System.out.println("------- CONTAS DO BANCO -------");
        for(Conta conta : this.contas){
            esperar(1800);
            conta.mostrarInfo();
        }
    }

    public void listarClientes(){
        System.out.println("------- CLIENTES DO BANCO -------");
        for(Cliente cliente : this.clientes){
            esperar(1800);
            System.out.println(cliente.dados());
            System.out.println();
        }
    }

    public List<Conta> getContas() {
        return this.contas;
    }

    public void setContas(List<Conta> contas) {
        this.contas = contas;
    }

    public List<Cliente> getClientes() {
        return this.clientes;
    }

    public void setClientes(List<Cliente> clientes) {
        this.clientes = clientes;
    }
    
}

File: Atividade5-TS/sistema-bancario/src/classes/Banco.ts
import Cliente from "./Cliente";
import Conta from "./Conta";
import Cadastro from "../utilidades/Cadastro";

import prompt from "prompt-sync";
let input = prompt();

export default class Banco {
  private _clientes: Cliente[] = inicializarClientes();
  private _contas: Conta[] = inicializarContas(this._clientes);

  public associarContaCliente(): void {
    let cliente: Cliente;
    let conta: Conta;
    let opcao: number = receberNumero("Já é cliente do banco? (1-sim/2-não): ");

    if (opcao == 1) {
      // se é cliente, procura na lista de clientes.
      cliente = this.procurarClientePeloCPF();
    } else {
      // se não é cliente, faz o cadastro do cliente.
      cliente = this.cadastrarCliente();
    }

    conta = Cadastro.criarConta(this._contas, cliente);
    this._contas.push(conta);
    cliente.adicionarConta(conta);
    msgSucesso();
  }

  public cadastrarCliente(): Cliente {
    let cliente: Cliente = Cadastro.cadastrarCliente(this._clientes);
    this._clientes.push(cliente);

    return cliente;
  }

  public fazerDeposito(): void {
    let conta: Conta = this.procurarContaPeloNumero();

    let valor: number = receberNumero("Valor do depósito: ");
    conta.depositar(valor);
    msgSucesso();
  }

  public fazerSaque(): void {
    let conta: Conta = this.procurarContaPeloNumero();

    let valor: number = receberNumero("Valor do saque: ");
    conta.sacar(valor);
    msgSucesso();
  }

  public fazerTransferencia(): void {
    let contaOrigem: Conta = this.procurarContaPeloNumero();
    let contaDestino: Conta = this.procurarContaPeloNumero();

    let valor: number = receberNumero("Valor a tranferir: ");
    contaOrigem.transferir(contaDestino, valor);
    msgSucesso();
  }

  public consultarConta(): void {
    let conta: Conta = this.procurarContaPeloNumero();
    conta.mostrarInfo();
  }

  public consultarCliente(): void {
    this.procurarClientePeloCPF();
  }

  public excluirConta(): void {
    let conta: Conta = this.procurarContaPeloNumero();

    let indice: number = this.indiceDaConta(conta.numero);
    this._contas.splice(indice, 1);

    let cliente: Cliente = conta.titular;
    let indiceConta: number = cliente.contas.findIndex(
      (contaCliente: Conta) => contaCliente.numero == conta.numero
    );
    cliente.contas.splice(indiceConta, 1);

    msgSucesso();
  }

  public excluirCliente(): void {
    let cliente: Cliente = this.procurarClientePeloCPF();

    let indice: number = this.indiceDoCliente(cliente.cpf);
    this._clientes.splice(indice, 1);

    for (let conta of cliente.contas) {
      let indiceConta: number = this.indiceDaConta(conta.numero);
      this._contas.splice(indiceConta, 1);
    }

    msgSucesso();
  }

  public listarContasCliente(): void {
    let cliente: Cliente = this.procurarClientePeloCPF();

    console.log(`\nContas do(a) cliente ${cliente.nome}`);
    cliente.contas.forEach((conta) => conta.mostrarInfo());
  }

  public totalizarSaldoCliente(): void {
    let total: number = 0;
    let cliente: Cliente = this.procurarClientePeloCPF();

    for (let conta of cliente.contas) {
      conta.mostrarInfo();
      total += conta.saldo;
    }

    console.log(`\nSaldo total de todas as contas: R$ ${total.toFixed(2)}`);
  }

  public totalAplicadoCliente() {
    let totalAplicado: number = 0;
    let cliente: Cliente = this.procurarClientePeloCPF();

    cliente.contas.forEach((conta) => (totalAplicado += conta.saldo));

    console.log(
      `\nTotal aplicado pelo(a) cliente ${
        cliente.nome
      }: R$ ${totalAplicado.toFixed(2)}`
    );
  }

  public realizarOrdemBancaria(): void {
    let contaOrigem: Conta = this.procurarContaPeloNumero();
    let contasDestino: Conta[] = this.receberContasDestino();

    for (let conta of contasDestino) {
      let valor: number = receberNumero(
        `Valor a ser transferido a conta ${conta.numero}: `
      );
      contaOrigem.transferir(conta, valor);
    }
    msgSucesso();
  }

  public receberContasDestino(): Conta[] {
    let qtdContas: number = receberNumero(
      "Quantidade de contas que quer transferir: "
    );
    let contasDestino: Conta[] = [];

    for (let i: number = 0; i < qtdContas; i++) {
      let conta: Conta = this.procurarContaPeloNumero();

      contasDestino.push(conta);
    }

    return contasDestino;
  }

  public transferirTitularidade(): void {
    let conta: Conta = this.procurarContaPeloNumero();
    let opcao = receberNumero(
      "\nO novo titular é cliente do banco? (1-sim/2-não): "
    );
    let novoTitular: Cliente;

    if (opcao == 2) {
      // novo titular não é cliente do banco, cadastra primeiro
      novoTitular = Cadastro.cadastrarCliente(this._clientes);
      this._clientes.push(novoTitular);
    } else {
      novoTitular = this.procurarClientePeloCPF();
    }

    // remove a conta da lista de contas do cliente antigo:
    let clienteAntigo: Cliente = conta.titular;
    let indiceConta: number = clienteAntigo.contas.findIndex(
      (contaCliente) => contaCliente.numero == conta.numero
    );
    clienteAntigo.contas.splice(indiceConta, 1);

    conta.titular = novoTitular;
    novoTitular.adicionarConta(conta);

    msgSucesso();
  }

  public obterTotalDinheiroDepositadoBanco(): number {
    let dinheiroTotal: number = 0;

    for (let conta of this._contas) {
      dinheiroTotal += conta.saldo;
    }

    return dinheiroTotal;
  }

  public mediaSaldoContasBanco(): void {
    let mediaSaldo: number =
      this.obterTotalDinheiroDepositadoBanco() / this._contas.length;

    console.log(
      `\nMédia do saldo de todas as contas do banco: R$ ${mediaSaldo.toFixed(
        2
      )}`
    );
  }

  public listarContasBanco(): void {
    console.log("\nContas do banco:");
    this._contas.forEach((conta) => conta.mostrarInfo());
  }

  public listarClientesBanco(): void {
    console.log("\nClientes do banco:");
    this._clientes.forEach((cliente) => console.log(cliente.mostrarDados()));
  }

  private procurarContaPeloNumero(): Conta {
    let numero: string = input("\nNúmero da conta: ");
    let conta: Conta | undefined = this._contas.find(
      (conta) => conta.numero === numero
    );

    if (!conta) {
      console.log("Conta não encontrada. Tente novamente:");
      return this.procurarContaPeloNumero();
    }

    console.log("\nConta encontrada!");
    conta.mostrarInfo();

    return conta;
  }

  private procurarClientePeloCPF(): Cliente {
    let CPF = input("\nCPF do cliente: ");
    let cliente: Cliente | undefined = this._clientes.find(
      (cliente) => cliente.cpf == CPF
    );

    if (!cliente) {
      console.log("Cliente não encontrado. Tente novamente: ");
      return this.procurarClientePeloCPF();
    }

    console.log(`\nCliente encontrado(a)!\n${cliente.mostrarDados()}`);

    return cliente;
  }

  private indiceDaConta(numero: string): number {
    return this._contas.findIndex((conta) => conta.numero == numero);
  }

  private indiceDoCliente(CPF: string): number {
    return this._clientes.findIndex((cliente) => cliente.cpf == CPF);
  }

  public get clientes(): Cliente[] {
    return this._clientes;
  }

  public set clientes(clientes: Cliente[]) {
    this._clientes = clientes;
  }

  public get contas(): Conta[] {
    return this._contas;
  }

  public set contas(contas: Conta[]) {
    this._contas = contas;
  }
}

import {
  inicializarClientes,
  inicializarContas,
  msgSucesso,
  receberNumero,
} from "../utilidades/Utils";
